File: src/lc/p2/py/add_two_number.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        ret = ListNode(0)
        
        carry = 0
        p = ret

        while True:
            if l1.next is None: #assume l1 and l2 are equal
                break
            l1 = l1.next
            l2 = l2.next

            sum_val = l1.val + l2.val + carry
            carry = int(sum_val/10)
            
            p.next = ListNode(sum_val % 10)
            p = p.next
            
        return ret.next

File: src/lc/p2/cpp/addTwoNumber.cpp
#include<iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode * ret = new ListNode(0);
        ListNode * p = ret;
        int e1, e2, carry(0), s;

        while(l1 != NULL || l2 != NULL || carry>0){
            p -> next = new ListNode(0);
            p = p -> next;

            e1 = (l1 == NULL)? 0 : l1 -> val;
            e2 = (l2 == NULL)? 0 : l2 -> val;
            s = e1 + e2 + carry;
            p->val = s%10;
            carry = (int)(s/10);

            l1 = (l1 == NULL)? NULL: l1 -> next;
            l2 = (l2 == NULL)? NULL: l2 -> next;
        }

        p = ret;
        ret = ret -> next;
        delete p;
        return ret;
    }
};